### PR TITLE
controller: Descriptive error on duplicate app.

### LINKED ISSUE #322: Creating duplicate apps causes a 500
```
$ flynn create example
Created example
$ flynn create example
POST http://dev.localflynn.com:443/apps: controller: unexpected status 500
```

This should print a validation error message instead.

### PR DESCRIPTION
**Pull request by archSeer**
Fixes #322.

Signed-off-by: Blaž Hrastnik <blaz.hrast@gmail.com>

### CODE REVIEW

File: controller/app.go [hunk 1]
```diff
@@ -55,6 +56,12 @@ func (r *AppRepo) Add(data interface{}) error {
 	}
 	meta := metaToHstore(app.Meta)
 	if err := r.db.QueryRow("INSERT INTO apps (app_id, name, meta, strategy) VALUES ($1, $2, $3, $4) RETURNING created_at, updated_at", app.ID, app.Name, meta, app.Strategy).Scan(&app.CreatedAt, &app.UpdatedAt); err != nil {
+		if e, ok := err.(*pq.Error); ok && e.Code.Name() == "unique_violation" && e.Constraint == "apps_name_idx" {
+			return httphelper.JSONError{
+				Code:    httphelper.ObjectExistsError,
+				Message: "Application with that name already exists!",
```
Comment (titanous) [inline controller/app.go]: This error should include the name `%q` and not have an exclamation point

Comment (archSeer) [inline controller/app.go]: Updated.

**Comment by titanous** [comment]
LGTM

File: controller/app.go [hunk 2]
```diff
@@ -55,6 +56,12 @@ func (r *AppRepo) Add(data interface{}) error {
 	}
 	meta := metaToHstore(app.Meta)
 	if err := r.db.QueryRow("INSERT INTO apps (app_id, name, meta, strategy) VALUES ($1, $2, $3, $4) RETURNING created_at, updated_at", app.ID, app.Name, meta, app.Strategy).Scan(&app.CreatedAt, &app.UpdatedAt); err != nil {
+		if e, ok := err.(*pq.Error); ok && e.Code.Name() == "unique_violation" && e.Constraint == "apps_name_idx" {
+			return httphelper.JSONError{
+				Code:    httphelper.ObjectExistsError,
+				Message: fmt.Sprintf("Application with name '%s' already exists.", app.Name),
+			}
```
Comment (lmars) [inline controller/app.go]: I have just merged #1209 so you should rebase this and change to something like:

```go
return httphelper.ObjectExistsErr(fmt.Sprintf("application %q already exists", app.Name))
```

File: controller/app.go [hunk 3]
```diff
@@ -55,6 +56,12 @@ func (r *AppRepo) Add(data interface{}) error {
 	}
 	meta := metaToHstore(app.Meta)
 	if err := r.db.QueryRow("INSERT INTO apps (app_id, name, meta, strategy) VALUES ($1, $2, $3, $4) RETURNING created_at, updated_at", app.ID, app.Name, meta, app.Strategy).Scan(&app.CreatedAt, &app.UpdatedAt); err != nil {
+		if e, ok := err.(*pq.Error); ok && e.Code.Name() == "unique_violation" && e.Constraint == "apps_name_idx" {
```
Comment (lmars) [inline controller/app.go]: This code is used in a few places, I think we should refactor it to something like:

```go
if postgres.IsUniquenessError(err, "apps_name_idx") {
	...
}
```

File: pkg/postgres/postgres.go [hunk 4]
```diff
@@ -201,3 +201,14 @@ func CleanUUID(u string) string {
 func FormatUUID(s string) string {
 	return s[:8] + "-" + s[8:12] + "-" + s[12:16] + "-" + s[16:20] + "-" + s[20:]
 }
+
+func IsUniquenessError(err error, constraint string) bool {
+	if e, ok := err.(*pq.Error); ok && e.Code.Name() == "unique_violation" && e.Constraint == constraint {
```
Comment (titanous) [inline pkg/postgres/postgres.go]: I don' think this logic is correct if `constraint == ""`?

**Comment by archSeer** [comment]
Comments addressed.

File: pkg/postgres/postgres.go (hunk 4 above)
Comment (archSeer) [inline pkg/postgres/postgres.go]: `""` should match any constraint.

File: pkg/postgres/postgres.go [hunk 5]
```diff
@@ -201,3 +201,14 @@ func CleanUUID(u string) string {
...
 	return s[:8] + "-" + s[8:12] + "-" + s[12:16] + "-" + s[16:20] + "-" + s[20:]
 }
+
+func IsUniquenessError(err error, constraint string) bool {
+	if e, ok := err.(*pq.Error); ok && e.Code.Name() == "unique_violation" {
+		if constraint != "" {
+			return e.Constraint == constraint
+		} else {
+			return true
+		}
```
Comment (titanous) [inline pkg/postgres/postgres.go]:
```go
return constraint == "" || constraint == e.Constraint
```

**Comment by titanous** [comment]
LGTM

**Pull request merged**